source/ota_mqtt.c

```diff
@@ -876,29 +881,31 @@ OtaErr_t requestJob_Mqtt( const OtaAgentContext_t * pAgentCtx )
         NULL
     };
     char reqCounterString[ U32_MAX_LEN + 1 ];
+
     /* NULL-terminated list of payload parts */
-    /* NOTE: this must agree with pOtaGetNextJobMsgTemplate, do not add spaces, etc. */
-    const char * pPayloadParts[] =
-    {
-        "{\"clientToken\":\"",
-        NULL, /* Request counter string not available at compile time, initialized below. */
-        ":",
-        NULL, /* Thing Name not available at compile time, initialized below. */
-        "\"}",
-        NULL
-    };
 
     assert( pAgentCtx != NULL );
 
     /* Suppress warnings about unused variables. */
     ( void ) pOtaJobsGetNextTopicTemplate;
     ( void ) pOtaGetNextJobMsgTemplate;
 
-    pTopicParts[ 1 ] = ( const char * ) pAgentCtx->pThingName;
-    pPayloadParts[ 1 ] = reqCounterString;
-    pPayloadParts[ 3 ] = ( const char * ) pAgentCtx->pThingName;
+    /* Client token max length is 64. It is a combination of request counter (max 10 characters), a separator colon, and the ThingName. */
+    xThingNameLength = ( uint32_t ) strnlen( ( const char * ) pAgentCtx->pThingName, OTA_CLIENT_TOKEN_MAX_THINGNAME_LEN );
+
+    reqCounterStringLength = ( uint32_t ) stringBuilderUInt32Decimal( reqCounterString, sizeof( reqCounterString ), reqCounter );
 
-    ( void ) stringBuilderUInt32Decimal( reqCounterString, sizeof( reqCounterString ), reqCounter );
+    /* Assemble the string by copying the pieces into the buffer. This is done manually since we know the size of the thingname. */
+    strcat( &pMsg[ msgSize ], "{\"clientToken\":\"" );
+    msgSize = 17U;
+    strncpy( &pMsg[ msgSize ], reqCounterString, reqCounterStringLength );
+    msgSize += reqCounterStringLength;
+    strcat( &pMsg[ msgSize ], ":" );
+    msgSize++;
+    strncpy( &pMsg[ msgSize ], ( const char * ) pAgentCtx->pThingName, xThingNameLength );
+    msgSize += xThingNameLength;
+    strcat( &pMsg[ msgSize ], "\"}" );
```

Comment:
strncat or strncpy or a separate length check before hand is necessary.

Reply:
This isn't necessary as the message is guaranteed to fit in the buffer.